rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,121 @@
+---
+feature: inherit-as-list
+start-date: 2021-10-17
+author: Ryan Burns (@r-burns)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: @synthetica9, @infinisil, @kevincox 
+shepherd-leader: @kevincox 
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new Nix syntax `<attrset>.[ <attrnames> ]`,
+which constructs a list from the values of an attrset.
+
+The goal is to provide a similarly-terse but more principled alternative
+to the often-used `with <attrset>; [ <attrnames> ]`.
+
+# Motivation
+[motivation]: #motivation
+
+It is currently cumbersome to create a list from the values of an attrset.
+If one has an attrset `attrs` and wishes to create a list containing some of
+its values, one could naively write:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+To avoid typing `attrs` many times, one will typically use `with` instead:
+
+```nix
+with attrs; [ a b c ]
+```
+
+However, the `with` expression has many well-known drawbacks, such as
+unintuitive shadowing behavior [1][2], prevention of static scope checking [3][4],
+and reduced evaluation performance [3].
+
+* [1] https://github.com/NixOS/nix/issues/490
+* [2] https://github.com/NixOS/nix/issues/1361
+* [3] https://github.com/NixOS/nixpkgs/pull/101139
+* [4] https://nix.dev/anti-patterns/language#with-attrset-expression
+
+Nonetheless, Nix expression authors are subtly guided toward the `with` form
+because it is (or at least appears) simpler than any existing alternatives.
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+attrs.[ a b c ]
```

Comment:
What about:

```
pkgs.[ python pythonPackages.mypackage ]
```

Should that be possible?

Reply:
It should, to cover previous behavior
```
with pkgs; [ python pythonPackages.mypackage ];
```

Reply:
The current implementation only allows attrnames in the brackets, but adding support for this syntax seems logical. I think this behavior is pretty intuitive and does what one would expect, and covers the previous `with pkgs` use case well.

Reply:
I would say that «this is just sugar for `a.[x y]` becoming `[a.x a.y]`» justifies `pkgs.[pythonPackages.mypy]` but not really `pkgs.[(python.override {})]`.

But yes, probably needs more explicit language.

Reply:
I think that's to be expected.

`pkgs.[(python.override {})` -> `[(pkgs.python.override {})]`

Reply:
Well, should `pkgs.[(wrapCC gcc)]` work in your opinion? `let wrapCCnew = …; in pkgs.[(wrapCCnew gcc)]`? What should be a simple rule to reason about this syntax?

Reply:
Yes, now I see where the problem is))) I agree that this option is very confusing)

In my opinion, it is logical to check variables for existence

```
pkgs.[(wrapCC gcc)]

PSEUDO CODE
[ ((wrapCC || pkgs.wrapCC) (gcc || pkgs.gcc)) ]
```

This will work exactly as it does now with `with pkgs; [(wrapCC gcc) ]`, but not dump all variables to the internal scope. 

We can make it simpler and forbid such overrides in this construct

```
pkgs.[stylua ormolu] ++ [inputs.agenix.packages.${system}.agenix];
```

But I don't know how convenient that would be at all.

Reply:
For example:

```
let
  vars = { a = 10 };
  a = 20
in vars.[a]
```

in my opinion, it should return 20;

Reply:
But then it doesn't seem to make sense to replace `with pkgs`. Maybe it's better

> We can make it simpler and forbid such overrides in this construct
> 
> pkgs.[stylua ormolu] ++ [inputs.agenix.packages.${system}.agenix];

Reply:
> ```
> pkgs.[(wrapCC gcc)]
> 
> PSEUDO CODE
> [ ((wrapCC || pkgs.wrapCC) (gcc || pkgs.gcc)) ]
> ```
> 
> This will work exactly as it does now with `with pkgs; [(wrapCC gcc) ]`, but not dump all variables to the internal scope.
> 

What do you mean by `not dump all variables to the internal scope`? isn't that exactly whats happening with this approach?
It would mean that variables in the internal scope will not be shadowed by those from the `pkgs` set, but it will still bring them into scope.. For example in the presence of an undefined variable reference (due to e.g. a mistake by the programmer) the semantics change. What would previously cause a runtime error might now evaluate (and to something the programmer did not intend).

Reply:
I agree with [7c6f434c](https://github.com/7c6f434c). We need to keep the rule simple. I am completely in favour of `pkgs.[foo] == [pkgs.foo]` and I think it is reasonable to extend to `pkgs.[foo.bar] == [pkgs.foo.bar]` but as soon as we pass a simple set of attributes I am definitely against. For example is `pkgs.[(wrapCC gcc)]` equivalent to `pkgs.[(pkgs.wrapCC gcc)]`,  `pkgs.[(wrapCC pkgs.gcc)]` or  `pkgs.[(pkgs.wrapCC pkgs.gcc)]`? Sure, we can define a rule but it starts to get hard for a reader to remember.

> In my opinion, it is logical to check variables for existence ... This will work exactly as it does now with `with`

I disagree. Now I need to know what variables exist to know how an expression will behave. And adding new packages can break expressions. There is a reason why a lot of people dislike `with` and want to create alternatives such as this very construct. I think that doing lookup like `with` basically defeats the point of this syntax in the first place. If you want that continue to use `with pkgs; [ ... ]`.

In general I would recommend that we start with just the single-identifier case. We can always add `pkgs.[foo.bar]` support in a future RFC once we see more examples of real-world use.

Reply:
> I disagree. Now I need to know what variables exist to know how an expression will behave. And adding new packages can break expressions. There is a reason why a lot of people dislike with and want to create alternatives such as this very construct. I think that doing lookup like with basically defeats the point of this syntax in the first place. If you want that continue to use with pkgs; [ ... ].

I was just pondering this in the posts above and ended up coming to the same conclusion as you)

Reply:
I cannot resolve conversations, but I think this one can be marked as such.